### ISSUE TITLE
KV filter multi-character field/value splitters 

### ISSUE THREAD

**Issue by fcntl** [logstash-plugins/logstash-filter-kv#15]
For the KV filter the value_split/field_split options are characters that "...form a regex character class". This means they are only single characters.

Please turn this into a full REGEX and not a character class, as this would allow multi-character delimiters and avoid having to reserve 2 ASCII characters (only 128 of them) for them, and make the KV filter a lot less fragile.

People can still use character classes inside the REGEX if they really have multiple delimiters values for field and value splitting.

**Comment by berglh** [comment]
Impossible not to get keys with messed up names when two delimiter character are present. A full implementation on regex KV delimiter would help a lot in these cases.
`Class = "\242\031\t\"\000\000\0017\000\001\002\000\2172\r\260\000\000\000\000\0006\346p\203>\017\223\001\320\366=\363\033{\004\000\000\000\000\000\2276\033" `

**Comment by colinsurprenant** [comment]
confirmed. the `@field_split` and `@value_split` options strings are assumed to be 1 char and embedded into a regex `[xyz]` character class construct which won't work for multi-chars splitters. I am currently fixing this filter for another context but will try to include a fix for this too.

**Comment by colinsurprenant** [comment]
correctly crafting the regex for this is a bit more involved than I originally though  so I will not be able to include a quick fix for it in my recent work on this plugin.

**Comment by trikosuave** [comment]
is there a workaround to field_split ","? fields are actually split by ", " (space) but that will split on both commas and spaces. I have both spaces and "," (no space) in my value and its messing it all up. :)
I have "," set for field_split and trimkey => " ". One of our fields is URL which the value has commas in the query string, so i can't just gsub ",![ ]" with something as the URL won't work.
Thanks!

**Comment by fiunchinho** [comment]
Any progress on this?

**Comment by CorbMax** [comment]
Same issue here. I need "::" as value_split.
Any progress?

**Comment by kai-kzh** [comment]
Same issue, need parse format [k1]=v1,[k2]=v2,[k3]=v3 and use ,\s*[ as separate to have tolerance for comma in value.

**Comment by Mojster** [comment]
@kai-kzh:
Use first mutate filter, to replace ,\s*[ with single character.
And after that apply kv filter.

**Comment by kai-kzh** [comment]
@Mojster Thank you for your suggestion. The problem here is, no matter which single character we choose, it can not appear in field/value.

**Comment by gmoskovicz** [comment]
It would be great to support multi char so one can easily parse the following:

```
UserAgent: com.apple.appstored/1.0 iOS/10.3.3 model/iPhone8,1 hwp/s8000 build/14G60 (6; dt:120), Client: Web browser
```

Please note the comma in between the user agent. If we want to parse this using the KV filter, first we need to mutate the `,/s` with `|` and then parse it. However if with KV you can define a comma and space as the separators, this would work smoothly.

**Comment by abraxxa** [comment]
We'd need this too for Cisco Firepower Threat Defense log messages.

**Comment by colinsurprenant** [comment]
I spent a few hours trying to craft a regex that would work with multi-chars field and value separators and I failed. With the way the parsing is currently done in this filter, I don't believe there is an easy solution to this and I think a whole new parsing strategy will have to be written to support it. I am open to ideas.

**Comment by jordansissel** [comment]
@gmoskovicz your example is a nice one, but it's unclear how reliable we could make it. Restating your example:

```
UserAgent: com.apple.appstored/1.0 iOS/10.3.3 model/iPhone8,1 hwp/s8000 build/14G60 (6; dt:120), Client: Web browser
```

If the 'UserAgent' is a free-form value provided by some network client, I assume this client could easily (by accident or by intent) include a `foo: bar` similar style entry here which would break the parser, right?

For example:

```
UserAgent: com.apple.appstored/1.0 iOS/10.3.3 model/iPhone8,1 hwp/s8000 build/14G60 (6; dt: 120), Client: Web browser
```

(I introduced a space in `dt: 120`) and now a computer would read this as:

* UserAgent: `com.apple.appstored/1.0 iOS/10.3.3 model/iPhone8,1 hwp/s8000 build/14G60 (6;`
* dt: `120),`
* Client: `web browser`

**Comment by gmoskovicz** [comment]
@jordansissel why a `;\s` will be parsed, if we only define `,\s` as the pattern?

But yes, if we have a comma and a space in between the user agent then we'll have other problems. However, i think that then it's better to change the source logger to use another separator.

**Comment by jordansissel** [comment]
@colinsurprenant I'm open to changing how this plugin's implementation works.

> We'd need this too for Cisco Firepower Threat Defense log messages.

@abraxxa It would be helpful for us to evaluate this if you can provide sample logs. Without this, we could deliver a solution to this issue that satisfies what we believe are the requirements that simultaneously does not solve your issue.

**Comment by colinsurprenant** [comment]
@jordansissel my thought was to introduce either a global switch 
```
config :multi_split, :validate => :boolean, :default => false
```
or two separate switches 
```
config :multi_field_split, :validate => :boolean, :default => false
config :multi_value_split, :validate => :boolean, :default => false
```
which would change the interpretation of the `field_split` and `value_split` options to threat them as multi-characters instead of single-character character-class.

I thought I'd be able to craft a regex that could handle multi-character split (using negative lookahead constructs to replace the character classes, which is far from idea to start with, but hoped it could work within the same regex logic) but realized it actually did not correctly work after adding a few more specs. 

Having such a "multi" switch would allow to choose a different parsing strategy depending if multi-character spitting is wanted or not and keep the same strategy for single-character splitting.

**Comment by abraxxa** [comment]
I‘m against any switches if possible and not required for backcompat.
The two configuration options should just be treated as regexes instead of building one from it just like the translate filter does it.

**Comment by colinsurprenant** [comment]
@abraxxa we either have to introduce a switch or a new option string for multi-character splitting because the **current** semantic of the `field_split` and `value_split` option is to use the given string as a character-class definition, meaning a series of different possible single-character splitters. For example, if you specify **today** `value_split => "&#"` in your config it will consider both `&` and `#` as 2 possible single-character splitters.

So for backward compatibility we have to introduce either a switch or new split string option.

**Comment by abraxxa** [comment]
I know...
Or a new plugin to replace this one.

**Comment by jordansissel** [comment]
I wonder if we could provide a nicer abstraction for users? Here is what I am thinking:

A user could specify a sample key-value entry:

```
  kv {
    sample => "key: value key2: value2"
  }
```

In the above proposed example, this tells Logstash to expect a `key` then `: ` (colon, space)`, then `value`. and that there's a space between the value and the next key.

For prior art, Go has a similar idea for their date formatting https://golang.org/pkg/time/#Time.Format where you write the format by example instead of a specific syntax or configuration.

Thoughts on this?

**Comment by colinsurprenant** [comment]
@jordansissel regardless of the configuration syntax/style we should support the current quoting and brackets handling. 

I am wondering if practically using such a *sample* style configuration will bring that much more value/usability related the implementation complexity to produce the parser for a given sample?

**Comment by jordansissel** [comment]
@colinsurprenant agree on quoting/bracket support remaining.

re: sample style configuration - I am still thinking about it. It may make user experience worse and also make certain definitions harder/impossible to specify -- not ideal, but I'm still trying to process how it could work.

That said, I can help with the regexp stuff. It's possible we can use this as an opportunity to simplify the parser code ;)

**Comment by guyboertje** [comment]
It was [suggested by Mojster](https://github.com/logstash-plugins/logstash-filter-kv/issues/15#issuecomment-324282023) earlier and declined by @kai-kzh - this is all UTF-8 though and you have the whole of the unicode gamut to choose from, [an example](https://unicode-table.com/en/#25E6).

I gave this answer in discuss regarding multi character splitting - it allows the user to specify exactly how they want the gsub to work. I mentioned the caveat that the gsub needs to specify the key names, however if a pattern can be found in the key names then all the better.  

Perhaps add a gsub config setting?
```
input {
  generator {
    message => 'key1: val1; key2: val2; key3:  https://site/?g={......"...;  CLR  rv:11.0)"..}; key4: val4;'
    count => 1
  }
}

filter {
  mutate {
    gsub => ["[message]", ";\s*(?<key>key1|key2|key3|key4):\s*", '▪\k<key>◦']
  }
  kv {
     field_split => "▪"
     value_split => "◦"
     source => "message"
  }
}

output {
  stdout {
    codec => rubydebug {metadata => true}
  }
}

```
Result:
```
{
      "sequence" => 0,
          "key2" => "val2",
    "@timestamp" => 2018-02-01T15:05:37.582Z,
          "key3" => "https://site/?g={......\"...;  CLR  rv:11.0)\"..}",
          "key4" => "val4;",
      "@version" => "1",
          "host" => "Elastics-MacBook-Pro.local",
       "message" => "key1: val1▪key2◦val2▪key3◦https://site/?g={......\"...;  CLR  rv:11.0)\"..}▪key4◦val4;"
}
```

**Comment by colinsurprenant** [comment]
@guyboertje interesting suggestion! I am also pretty sure we can probably always pre-process a message using mutate/gsub prior to kv to make it work with single-char splitter. Using unicode is also an excellent idea.

But do we need to add an extra gsub option for that? not sure.

**Comment by guyboertje** [comment]
My logic for suggesting a gsub setting here is:
- It does not cause a separate filter invocation.
- The target for the gsub is defined already in `source` - no possible disconnect.
- The docs for the KV filter makes for better reading when describing this case (rather than saying use mutate/gsub)
- If a consultant builds a solution using mutate/gsub plus kv walks away, later the significance or coupling between these two filter sections may not be apparent but the intention to replace characters as part of the kv filter config would be.

**Comment by colinsurprenant** [comment]
@guyboertje OTOH the idea of separation of concerns through filters and composition using the config language is what LS is about. I understand your points but this reasoning could be applied to all configurations and by this logic we should write single do-it-all filters. 

From a performance perspective I'd like to see the difference of including `gsub` as an option versus having it done through the mutate filter. My feeling here is that if performance is of paramount importance, a custom filter tailored fit to the exact extraction pattern might ultimately be a better idea.

**Comment by fbaligand** [comment]
The gsub idea is nice but has a flaw : it requires to list all possible keys in gsub pattern.
And to me, this precisely the kv filter strength compared to a grok filter : you don't have to know the keys, filter detect them dynamically.

**Comment by guyboertje** [comment]
@fbaligand 
Sure it has flaws, but is it really possible to solve this flawlessly - for every edge case, no additional hints?

The `gsub` idea works if the keys have a unique pattern as well. Perhaps only some keys are known to be of interest in the rest of the config and Kibana.

IMO, the generalized definition of this is:
- A **key value** separator is some characters that appear directly after a key.
- A **value key** separator is some characters that appear directly before a key.

A flawless solution is therefore concerned with knowing what the keys are and the separators directly before and after them are, Anything else is a value.

If you have multiple character separators and you know that they do not appear in keys or values then you can use [Dissect indirect fields](https://www.elastic.co/guide/en/logstash/current/plugins-filters-dissect.html#_indirect_field_notation).

**Comment by colinsurprenant** [comment]
@yaauie has just contributed a great refactor to the regex builder #54 which will allow the introduction of  multi-character splitters. I rebooted my work on it on top of @yaauie's changes and submitted PR #55.

- I would like to add more multi-chars splitting examples to the specs: if you have specific examples of log line that could benefit multi-char splitters for fields or values then please submit them in the PR comments and I will add them. Please be specific about the example you submit, what is the splitter you suggest and the result you are seeking.

- I opted to introduce two new options `field_split_multi` and `value_split_multi` which are unset by default to preserve backward compatibility. Please comment in the PR if you have better suggestions about this.

**Comment by colinsurprenant** [comment]
@guyboertje @fbaligand FYI in #55 I was able to craft a regex to parse without mutate/gsub
```
key1: val1; key2: val2; key3:  https://site/?g={......"...;  CLR  rv:11.0)"..}; key4: val4;
```

**Comment by fbaligand** [comment]
Great @colinsurprenant !

**Comment by fbaligand** [comment]
By the way, your sample to test multichar field/value splitters is exactly the case I need to solve !

**Comment by colinsurprenant** [comment]
#55 is merged and plugin published as V4.1.0 🎉 

Thanks all for your PATIENCE and help! 😄

**Comment by fbaligand** [comment]
Great new @colinsurprenant !
Big Thanks for this enhancement !
